### PR TITLE
precalculate densities

### PR DESCRIPTION
**Pull request by evanmcc**
for hip17 scaling, precalculate all densities so that scale can be cheaply calculated.  2-3x speedup locally.

### CODE REVIEW

**Comment by mrallen1** [comment]
Instead of usort after the fold maybe use a map to quickly check if a hex is already included in the result accumulator during the foldl

**Comment by mrallen1** [comment]
Or just blindly add all hexes to a set

File: src/blockchain_hex.erl [hunk 1]
```diff
@@ -1,27 +1,20 @@
...
 
 -include("blockchain_vars.hrl").
 -include_lib("common_test/include/ct.hrl").
 
--define(SCALE_MEMO_TBL, '__blockchain_hex_scale_memoization_tbl').
--define(DENSITY_MEMO_TBL, '__blockchain_hex_density_memoization_tbl').
+-define(PRE_UNCLIP_TBL, '__blockchain_hex_unclipped_tbl').
+-define(PRE_CLIP_TBL, '__blockchain_hex_clipped_tbl').
+
 -define(ETS_OPTS, [named_table, public]).
```
Comment (vihu) [inline src/blockchain_hex.erl:15]: This can be removed now I think?

File: src/blockchain_hex.erl [hunk 2]
```diff
@@ -1,27 +1,20 @@
...
-         destroy_memoization/0]).
-
--ifdef(TEST).
--export([densities/3]).
--endif.
+         destroy_memoization/0
+        ]).
 
 -include("blockchain_vars.hrl").
 -include_lib("common_test/include/ct.hrl").
```
Comment (vihu) [inline src/blockchain_hex.erl:10]: Also can be removed?

File: src/blockchain_hex.erl [hunk 3]
```diff
@@ -121,214 +122,135 @@ var_map(Ledger) ->
...
-        case ets:lookup(TblName, Key) of
-            [{_Key, Res}] -> {ok, Res};
-            [] -> not_found
-        end
-    catch
-        %% if the table doesn't exist yet, create it and return `not_found'
-        error:badarg ->
-            _ = maybe_start(TblName),
-            not_found
-    end.
```
Comment (vihu) [inline src/blockchain_hex.erl:125]:
```suggestion
-spec ulookup(Key :: non_neg_integer()) -> non_neg_integer().
```

File: src/blockchain_hex.erl [hunk 4]
```diff
@@ -121,214 +122,135 @@ var_map(Ledger) ->
...
 
--spec maybe_start(TblName :: tblnames()) -> tblnames().
-maybe_start(TblName) ->
-    try
-        _TblName = ets:new(TblName, ?ETS_OPTS)
-    catch
-        error:badarg -> TblName
+ulookup(Key) ->
```
Comment (vihu) [inline src/blockchain_hex.erl:126]:
```suggestion
-spec ulookup(Key :: non_neg_integer()) -> non_neg_integer().
ulookup(Key) ->
```

File: src/blockchain_hex.erl [hunk 5]
```diff
@@ -121,214 +122,135 @@ var_map(Ledger) ->
...
-    H3Index :: h3:h3_index(),
-    VarMap :: var_map(),
-    Ledger :: blockchain_ledger_v1:ledger()
-) -> densities().
-densities(H3Index, VarMap, Ledger) ->
-    case lookup(H3Index, ?DENSITY_MEMO_TBL) of
-        {ok, Densities} -> Densities;
-        not_found -> memoize(?DENSITY_MEMO_TBL, H3Index,
-                            calculate_densities(H3Index, VarMap, Ledger))
+clookup(Key) ->
```
Comment (vihu) [inline src/blockchain_hex.erl:132]:
```suggestion
-spec clookup(Key :: non_neg_integer()) -> non_neg_integer().
clookup(Key) ->
```

File: src/blockchain_hex.erl [hunk 6]
```diff
@@ -121,214 +122,135 @@ var_map(Ledger) ->
...
-    H3Root :: h3:h3_index(),
-    VarMap :: var_map(),
-    Locations :: locations(),
-    Ledger :: blockchain_ledger_v1:ledger()
-) -> densities().
-densities(H3Root, VarMap, Locations, Ledger) ->
-    case maps:size(Locations) of
-        0 ->
-            {#{}, #{}};
+maybe_precalc(Ledger) ->
```
Comment (vihu) [inline src/blockchain_hex.erl:138]:
```suggestion
-spec maybe_precalc(Ledger :: blockchain_ledger_v1:ledger()) -> ok.
maybe_precalc(Ledger) ->
```

File: src/blockchain_hex.erl [hunk 7]
```diff
@@ -121,214 +122,135 @@ var_map(Ledger) ->
...
-    OccupiedHexesThisRes = maps:keys(UD),
-
-    DensityTarget = maps:get(tgt, maps:get(Res, VarMap)),
-
-    M1 = lists:foldl(
-        fun(ThisResHex, Acc3) ->
-            OccupiedCount = occupied_count(DensityTarget, ThisResHex, UD),
-            Limit = limit(Res, VarMap, OccupiedCount),
-            maps:put(ThisResHex, min(Limit, maps:get(ThisResHex, M0)), Acc3)
+precalc(Ledger) ->
```
Comment (vihu) [inline src/blockchain_hex.erl:146]:
```suggestion
-spec precalc(Ledger :: blockchain_ledger_v1:ledger()) -> ok.
precalc(Ledger) ->
```

File: src/blockchain_hex.erl [hunk 8]
```diff
@@ -121,214 +122,135 @@ var_map(Ledger) ->
...
-                         case blockchain_ledger_v1:find_gateway_last_challenge(GWAddr, Ledger) of
-                             {ok, undefined} -> false;
-                             {ok, LastChallenge} ->
-                                 (CurrentHeight - LastChallenge) =< InteractiveBlocks;
-                             {error, not_found} -> false
-                         end
-                    end, GWs).
+    ets:new(?PRE_UNCLIP_TBL, [named_table, public]),
+    ets:new(?PRE_CLIP_TBL, [named_table, public]),
+
```
Comment (vihu) [inline src/blockchain_hex.erl:157]: nit: too much whitespace

File: src/blockchain_hex.erl [hunk 9]
```diff
@@ -121,214 +122,135 @@ var_map(Ledger) ->
...
     {ok, CurrentHeight} = blockchain_ledger_v1:current_height(Ledger),
-    lists:filter(fun(GWAddr) ->
-                         case blockchain_ledger_v1:find_gateway_last_challenge(GWAddr, Ledger) of
-                             {ok, undefined} -> false;
-                             {ok, LastChallenge} ->
-                                 (CurrentHeight - LastChallenge) =< InteractiveBlocks;
-                             {error, not_found} -> false
-                         end
-                    end, GWs).
+    ets:new(?PRE_UNCLIP_TBL, [named_table, public]),
```
Comment (vihu) [inline src/blockchain_hex.erl:155]: Actually use the `ETS_OPTS` here :slightly_smiling_face:

File: src/blockchain_hex.erl (hunk 7 above)
Comment (vihu) [reply src/blockchain_hex.erl:146]: Is it possible we can split this giant function a little bit, I understand it was horrible to begin with but would be good to separate it out into slightly logically named functions if possible at all

**Review by vihu — APPROVED** [review]

**Pull request merged**